### PR TITLE
Preload Post Purchase Upsell experiment assignment

### PR DESCRIPTION
**Pull request by razvanpapadopol**
Related to pbxNRc-2eP-p2#comment-4003

#### Proposed Changes

* When adding a premium plan to cart, preload the 20989-explat-experiment assignment.

#### Testing Instructions

* Add a Premium plan to cart
* When the checkout page is loaded, a network request to the URL: `public-api.wordpress.com/wpcom/v2/experiments/0.1.0/assignments/calypso?_envelope=1&experiment_name=calypso_plans_postpurchasepage_newdesign_202301_v1` should be made. It returns an object with the correct experiment variation.

#### Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

Related to #

### CODE REVIEW

File: client/my-sites/checkout/composite-checkout/components/wp-checkout.tsx
Comment (sirbrillig) [inline client/my-sites/checkout/composite-checkout/components/wp-checkout.tsx]: I don't know enough to be sure; is it safe to call `loadExperimentAssignment()` on every render?

**Review by sirbrillig — COMMENTED** [review]
I didn't run through the tests (I leave that to your team) but from a checkout perspective this should be fine as long as loadExperimentAssignment is idempotent.

**Review by paulopmt1 — APPROVED** [review]
It's working great @razvanpapadopol! I tested the experiment page load with and without this code (coming from the /checkout page) and saw an average of 400ms difference between when we click the checkout button and when we see the upsell screen.


We're storing the data properly:

image

Profile with cache:

experiment with-cache


Profile without cache:

image

**Comment by paulopmt1** [comment]
Sorry for the notification bomb folks, apparently I still didn't get the correct rebase workflow. Next time will use `--force-with-lease ` with the `push` command (I though it was only necessary when using the `force` parameter)

File: client/my-sites/checkout/composite-checkout/components/wp-checkout.tsx
Comment (razvanpapadopol) [reply client/my-sites/checkout/composite-checkout/components/wp-checkout.tsx]: Good point, @sirbrillig. I've investigated a bit and it looks ok. As it's mentioned also in the [package readme](https://github.com/Automattic/wp-calypso/tree/trunk/packages/explat-client#usage):
> Call as many times and as much as you like, we manage the state and requests.

So even if the preloading function was designed to be used outside of the React application (signup controller was the first usage), it should be safe to call it on multiple renders.

But in a React Component any side effect should be explicitly defined, together with its dependencies. So I've wrapped the call in a `useEffect` and placed it before any of the component returns (https://github.com/Automattic/wp-calypso/pull/71899/commits/d7209c2721d4456bec6f3c2589a048f36b94f26e). There shouldn't be any noticeable change in behavior. The function will still get called multiple times when a new `responseCart` object is passed.

**Review by DustyReagan — APPROVED** [review]
I see the network request as well. Seems to be working! 👍

**Comment by paulopmt1** [comment]
Before merging this, we'll need to configure the ongoing experiment to use triggers. Let me know when you feel it's good to go @razvanpapadopol so I'll go ahead and change the experiment + deploying this improvement.

Thanks!

**Comment by paulopmt1** [comment]
I'll close this now, so we'll use it in follow-up experiments. Here's more context: p1673541619155519/1673289198.965569-slack-CKZHG0QCR

Thanks for all your work on it!

**Pull request closed without merging**
